### PR TITLE
feat: Multienv support for Lambda layers

### PR DESCRIPTION
**Pull request by jhockett**
*Issue #, if available:*
#4886 #5005 #5174

*Description of changes:*
New Lambda layers are now differentiated by environment
layerVersionMap is now apart of the `team-provider-info.json` file, runtimes are stored in the layer directory in `layer-runtimes.json`
Layer CFN file makes better use of Refs

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerHelpers.ts [hunk 1]
```diff
@@ -155,6 +157,56 @@ export function previousPermissionsQuestion(layerName: string): ListQuestion[] {
...
+      _.has(teamProviderInfo, [env, 'nonCFNdata', categoryName, layerName, 'runtimes']) &&
+      _.has(teamProviderInfo, [env, 'nonCFNdata', categoryName, layerName, 'layerVersionMap']),
+  );
+  const currentEnv = context.amplify.getEnvInfo().envName;
+  if (filteredEnvs.includes(currentEnv)) {
+    return _.get(teamProviderInfo, [currentEnv, 'nonCFNdata', categoryName, layerName]);
+  }
+  context.print.info(`Adding Lambda layer ${layerName} to ${currentEnv} environment.`);
+  const { envName } = await prompt(chooseParamsOnEnvInitQuestion(layerName, filteredEnvs));
+  const defaultPermission = [{ type: 'private' }];
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerHelpers.ts:174]: `deafaultPermission` gets referenced more than once. I'm going to leave as is.

**Comment by r0zar** [comment]
Tested this locally with a basic project. Looking good! :heavy_check_mark: 

https://github.com/r0zar/amplify-layers-test

Whats the recommended approach to keep the `team-provider-info.json` small?

If a project has frequent changes to layers, the versions might start to bloat the file.

Thanks for this!

**Comment by jhockett** [comment]
> Tested this locally with a basic project. Looking good! ✔️
> 
> https://github.com/r0zar/amplify-layers-test
> 
> Whats the recommended approach to keep the `team-provider-info.json` small?
> 
> If a project has frequent changes to layers, the versions might start to bloat the file.
> 
> Thanks for this!

Thanks for the testing and feedback @r0zar! There isn't currently a great solution to minimizing the size of the `team-provider-info` file. In the future, we plan to add functionality to delete individual layer versions.

File: packages/amplify-e2e-core/src/categories/lambda-layer.ts [hunk 2]
```diff
@@ -1,11 +1,12 @@
+import { JSONUtilities } from 'amplify-cli-core';
 import fs from 'fs-extra';
 import _ from 'lodash';
 import path from 'path';
 import { nspawn as spawn, ExecutionContext, getCLIPath, KEY_DOWN_ARROW } from '../../src';
 import { getLayerVersion, listVersions } from '../utils/sdk-calls';
 import { multiSelect } from '../utils/selectors';
 
-type LayerRuntimes = 'dotnetcore3.1' | 'go1.x' | 'java' | 'nodejs' | 'python';
+export type LayerRuntimes = 'nodejs' | 'python';
```
Comment (edwardfoyle) [inline packages/amplify-e2e-core/src/categories/lambda-layer.ts:9]: Are we removing official support for using LL with those other runtimes? If so, do we have a docs PR for that limitation?

File: packages/amplify-e2e-core/src/init/amplifyPull.ts [hunk 3]
```diff
@@ -1,15 +1,41 @@
...
   return new Promise((resolve, reject) => {
     const tableHeaderRegex = /\|\sCategory\s+\|\sResource\sname\s+\|\sOperation\s+\|\sProvider\splugin\s+\|/;
     const tableSeperator = /\|(\s-+\s\|){4}/;
 
-    const chain = spawn(getCLIPath(), ['pull'], { cwd, stripColors: true })
-      .wait('Pre-pull status')
-      .wait('Current Environment')
-      .wait(tableHeaderRegex)
-      .wait(tableSeperator);
+    const chain = spawn(getCLIPath(), settings.appId ? ['pull', '--appId', settings.appId] : ['pull'], { cwd, stripColors: true });
```
Comment (edwardfoyle) [inline packages/amplify-e2e-core/src/init/amplifyPull.ts:8]: are these changes related to LL?

File: packages/amplify-provider-awscloudformation/lib/initialize-env.js [hunk 4]
```diff
@@ -47,8 +47,7 @@ function run(context, providerMetadata) {
         });
       });
 
-      const jsonString = JSON.stringify(amplifyMeta, null, 4);
-      fs.writeFileSync(amplifyMetafilePath, jsonString, 'utf8');
+      context.amplify.writeObjectAsJson(amplifyMetafilePath, amplifyMeta, true);
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/lib/initialize-env.js:50]: if it's not too much trouble, would be nice to use JSONUtilities here and in the placed you modified in push-resources. But we could also come back and refactor these files later.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerParams.ts [hunk 5]
```diff
@@ -18,7 +20,7 @@ export type LayerParameters = {
   build: boolean;
 };
 
-export type StoredLayerParameters = Pick<LayerParameters, 'runtimes' | 'layerVersionMap'>;
+export type StoredLayerParameters = Pick<Partial<LayerParameters>, 'runtimes' | 'layerVersionMap'>;
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerParams.ts:23]: StoredLayerParameters do not include a runtimes field for new layers. They're optional in the sense that old style layers will have them, but new ones will not.

File: packages/amplify-e2e-core/src/categories/lambda-layer.ts (hunk 2 above)
Comment (jhockett) [reply packages/amplify-e2e-core/src/categories/lambda-layer.ts:9]: We've only ever supported nodejs and python.

File: packages/amplify-e2e-core/src/init/amplifyPull.ts (hunk 3 above)
Comment (jhockett) [reply packages/amplify-e2e-core/src/init/amplifyPull.ts:8]: It's a generic upgrade to the amplifyPull e2e helper I wrote previously, but it's now also used in a LL e2e test

File: packages/amplify-category-function/src/provider-utils/awscloudformation/index.ts [hunk 6]
```diff
@@ -306,31 +308,51 @@ function getHeadlessParams(context, resourceName) {
...
+      return functionParams;
+    }
 
-  if (resourceParams.trigger === true) {
-    envParams = await initTriggerEnvs(context, resourceParams, providerPlugin, envParams, srvcMetaData);
+    if (resourceParams.trigger === true) {
+      envParams = await initTriggerEnvs(context, resourceParams, providerPlugin, envParams, srvcMetaData);
+    }
+    return envParams;
+  } else if (isMultiEnvLayer(context, resourceName) && service === ServiceName.LambdaLayer) {
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/index.ts:337]: Since we did not previously, I decided to leave non-multiEnv layer functionality as is.

File: packages/amplify-provider-awscloudformation/lib/initialize-env.js (hunk 4 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/lib/initialize-env.js:50]: I'm going to leave as is for now. Each file deserves a separate thorough refactor to use `JSONUtilities` and `StateManager` from cli core.

**Comment by kianryan** [comment]
Having some problems building this branch on WSL2:

``
lerna ERR! yarn run build exited 1 in 'amplify-category-function'
lerna ERR! yarn run build stdout:
$ tsc
src/index.ts:5:10 - error TS2305: Module '"../node_modules/amplify-cli-core/lib"' has no exported member 'stateManager'.

5 import { stateManager } from 'amplify-cli-core';
           ~~~~~~~~~~~~
``

**Comment by jhockett** [comment]
> Having some problems building this branch on WSL2:
> 
> ``
> lerna ERR! yarn run build exited 1 in 'amplify-category-function'
> lerna ERR! yarn run build stdout:
> $ tsc
> src/index.ts:5:10 - error TS2305: Module '"../node_modules/amplify-cli-core/lib"' has no exported member 'stateManager'.
> 
> 5 import { stateManager } from 'amplify-cli-core';
> ~~~~~~~~~~~~
> ``

Fixed

**Comment by kianryan** [comment]
Managed to get the build working, many thanks.

When switching between environments, the cli correctly tracks layers to the environment so changes to the layer are correctly differentiated by the current environment.  However in the cloudformation template for lambda functions, the layer reference remains pointing to the environment the function was last updated in.  The layer/environment reference is only updated when the layer reference is updated via cli.

Is this an intentional decision?  This currently means that after a environment checkout, each function requires updating to point to the current environment.  Can see reasons for /not/ changing this on checkout, but if possible, we would like some clarification.

**Comment by jhockett** [comment]
> However in the cloudformation template for lambda functions, the layer reference remains pointing to the environment the function was last updated in. The layer/environment reference is only updated when the layer reference is updated via cli.
> 
> Is this an intentional decision? This currently means that after a environment checkout, each function requires updating to point to the current environment.

Thanks for the feedback @kianryan! It's intentional in the sense that it was out of scope for this PR, but this is something we can improve on in the future. I'll make sure to bring this up with the team.

**Comment by kianryan** [comment]
> > However in the cloudformation template for lambda functions, the layer reference remains pointing to the environment the function was last updated in. The layer/environment reference is only updated when the layer reference is updated via cli.
> > 
> > Is this an intentional decision? This currently means that after a environment checkout, each function requires updating to point to the current environment.
> 
> Thanks for the feedback @kianryan! It's intentional in the sense that it was out of scope for this PR, but this is something we can improve on in the future. I'll make sure to bring this up with the team.
> 

Okay, we can work with that. Thank you for the prompt response.

**Comment by chunyenHuang** [comment]
Thanks for the work!

**Comment by kennandavison** [comment]
How do I generate layer-runtimes.json? I don't have one.
